Sources/_StringProcessing/Regex/DSLTree.swift

```diff
@@ -583,16 +589,28 @@ extension DSLTree.Node {
 
   /// Returns the type of the whole match, i.e. `.0` element type of the output.
   var wholeMatchType: Any.Type {
-    if case .matcher(let type, _) = outputDefiningNode {
+    switch outputDefiningNode {
+    case .matcher(let type, _):
       return type
+    case .typeErase:
+      return AnyRegexOutput.self
+    default:
+      return Substring.self
     }
-    return Substring.self
   }
 }
 
 extension DSLTree {
+  @available(SwiftStdlib 5.7, *)
   var captureList: CaptureList {
     var list = CaptureList()
+    // FIXME: This is peering through any top-level `.typeErase`. Once type
+    // erasure was handled in the engine, this can be simplified to using `root`
+    // directly.
+    var root = root
+    while case let .typeErase(child) = root {
+      root = child
+    }
```

Comment:
Why is this scan in here? It seems like you'd want to capture an `ARO`, is that not possible?